oras/provider.py

```diff
@@ -893,6 +893,7 @@ def do_request(
         :type stream: bool
         """
         headers = headers or {}
+        headers.update(self.headers)
```

Comment:
This should likely be tried after an initial failure (e.g., see below where it updates after the auth response.

I think the maintain the current behavior we might want to have a boolean that determines if the headers should be re-used and persist (or not).

Reply:
Thanks for quick  reply!

After initial failure it tries to get token from realm returned in 'www-authenticate' header, but remote registry does not expose it externally, and all the following re-auth attempts fail since URL 'https://docker-registry-realm' is not available.

We can pass refresh_headers flag with False value to keep self.headers in request, like it is done in the pull method. But my concerns here are:
- it is necessary to pass it through long chain (push -> upload_blob -> put_upload/chunked_upload);
- if user performed implicit login/set_basic_auth operation, then he meant to use the credentials until he calls logout operation

What do you think?

Reply:
The normal flow I think is to expect that auth flow to work, although now that I think of it the auth isn't standardized. The basic auth is usually there to then request the token (the flow we have).

I understand your use case I think, and perhaps we need another set credential function that sets something permanently and does not try the auth flow?

Reply:
If you have a set method for more static credentials, that would add those credentials to the headers, and given the client knows that set method was used, it would not touch or try to update headers.

Reply:
That was my initial point, basic auth (username and pass) are static unlike the token and could be used in all requests until the explicit logout (like in ORAS CLI).

Reply:
But that function is used to encode in the initial request for the token. Saying to take the basic auth and not use it for a token (just provide as is) is a different flow I think.

Reply:
I see. Fixed in following commit to propagate refresh_headers flag like in pull flow.

I guess, refresh_headers flag may be turned later to the internal property, and set automatically within set_token_auth method to True and within the set_basic_auth method to False.